lib/rule_engine/ast.py

```diff
@@ -377,14 +381,28 @@ def __op_arithmetic(self, op, thing):
 		_assert_is_numeric(right_value)
 		return op(left_value, right_value)
 
-	_op_add  = functools.partialmethod(__op_arithmetic, operator.add)
 	_op_sub  = functools.partialmethod(__op_arithmetic, operator.sub)
 	_op_fdiv = functools.partialmethod(__op_arithmetic, operator.floordiv)
 	_op_tdiv = functools.partialmethod(__op_arithmetic, operator.truediv)
 	_op_mod  = functools.partialmethod(__op_arithmetic, operator.mod)
 	_op_mul  = functools.partialmethod(__op_arithmetic, operator.mul)
 	_op_pow  = functools.partialmethod(__op_arithmetic, operator.pow)
 
+class AddExpression(LeftOperatorRightExpressionBase):
+	"""A class for representing arithmetic expressions from the grammar text such as addition and subtraction."""
+	compatible_types = (DataType.FLOAT,DataType.STRING)
+	result_type =  DataType.STRING
+	def __op_add(self, op, thing):
+		left_value = self.left.evaluate(thing)
+		right_value = self.right.evaluate(thing)
+		if isinstance(left_value,str) or isinstance(right_value, str):
+			_assert_is_string(left_value)
+			_assert_is_string(right_value)
+		return op(left_value, right_value)
```

Comment:
ah wait, missed the case of validating if one is not a str then both should be float. will make a revision